Drop IsDeprecatedWeakRefSmartPointerException from WebMDNSRegister

RupinMittal opened the pull request:
#### f76b826b5fc047c430744a474795de677f948f22
<pre>
Drop IsDeprecatedWeakRefSmartPointerException from WebMDNSRegister
<a href="https://bugs.webkit.org/show_bug.cgi?id=281729">https://bugs.webkit.org/show_bug.cgi?id=281729</a>
<a href="https://rdar.apple.com/138170193">rdar://138170193</a>

Reviewed by Chris Dumez.

WebMDNSRegister forwards its ref-counting to its owner LibWebRTCNetwork,
which forwards its own ref-counting to its owner WebProcess.

* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetwork.cpp:
(WebKit::LibWebRTCNetwork::LibWebRTCNetwork):
(WebKit::LibWebRTCNetwork::ref const):
(WebKit::LibWebRTCNetwork::deref const):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetwork.h:
(WebKit::LibWebRTCNetwork::create): Deleted.
(WebKit::LibWebRTCNetwork::connection): Deleted.
(WebKit::LibWebRTCNetwork::isActive const): Deleted.
(WebKit::LibWebRTCNetwork::monitor): Deleted.
(WebKit::LibWebRTCNetwork::socketFactory): Deleted.
(WebKit::LibWebRTCNetwork::disableNonLocalhostConnections): Deleted.
(WebKit::LibWebRTCNetwork::resolver): Deleted.
(WebKit::LibWebRTCNetwork::mdnsRegister): Deleted.
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.cpp:
(WebKit::LibWebRTCNetworkManager::unregisterMDNSNames):
(WebKit::LibWebRTCNetworkManager::CreateNameForAddress):
* Source/WebKit/WebProcess/Network/webrtc/WebMDNSRegister.cpp:
(WebKit::WebMDNSRegister::WebMDNSRegister):
(WebKit::WebMDNSRegister::ref const):
(WebKit::WebMDNSRegister::deref const):
(WebKit::WebMDNSRegister::registerMDNSName):
* Source/WebKit/WebProcess/Network/webrtc/WebMDNSRegister.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::libWebRTCNetwork):
(WebKit::WebProcess::protectedLibWebRTCNetwork):
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: <a href="https://commits.webkit.org/285554@main">https://commits.webkit.org/285554@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0dd9253e23c89918a00739b89fcc673c063c5251

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72907 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52332 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25707 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77109 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24141 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60137 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23957 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57344 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15817 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75974 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47307 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62750 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37746 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43952 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22474 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65809 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20574 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78778 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/17153 "Built successfully") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19723 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Running apply-patch; Checked out pull request; Running run-layout-tests-in-stress-mode; layout-tests running") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65777 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/17202 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62756 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65060 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13362 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7022 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11239 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/48130 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2917 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/49197 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50492 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48942 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetwork.h:
```diff
@@ -92,7 +93,7 @@ class LibWebRTCNetwork : private FunctionDispatcher, private IPC::MessageReceive
 #endif
 #if ENABLE(WEB_RTC)
-    WebMDNSRegister m_mdnsRegister;
+    Ref<WebMDNSRegister> m_mdnsRegister;
```
Not sure if this is correct since ideally it would not be a `Ref<>` and `WebMDNSRegister` would just forward ref/deref to `LibWebRTCNetwork`. I actually did try to make `LibWebRTCNetwork` ref-counted, but ran into an issue. It's owned by `WebProcess`, so I tried to make `LibWebRTCNetwork` forward its own ref-counting to `WebProcess`. But `WebProcess` isn't ref-counted. Should it be? Or should I just make `LibWebRTCNetwork` inherit from `RefCounted<>`?. What's the best course of action? Or perhaps this patch is fine as is?

geoffreygaren replied:
`WebProcess` is a singleton. You could make it empty refcounted, I guess.

cdumez replied:
Indeed, adding empty ref() / deref() to WebProcess and forward ref-counting.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetwork.h:
```diff
@@ -62,12 +62,13 @@ class LibWebRTCNetwork : private FunctionDispatcher, private IPC::MessageReceive
 
 #if ENABLE(WEB_RTC)
     WebMDNSRegister& mdnsRegister() { return m_mdnsRegister; }
+    Ref<WebMDNSRegister> protectedmdnsRegister() { return m_mdnsRegister; }
```
`protectedMDNSRegister()`

cdumez commented on Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetwork.h:
```diff
@@ -35,15 +35,20 @@
     WTF_MAKE_TZONE_ALLOCATED(LibWebRTCNetwork);
 public:
-    static UniqueRef<LibWebRTCNetwork> create() { return UniqueRef { *new LibWebRTCNetwork() }; }
+    static UniqueRef<LibWebRTCNetwork> create(WebProcess& webProcess) { return UniqueRef { *new LibWebRTCNetwork(webProcess) }; }
```
Would be surprising if this built without updating the caller.

Also, it would be better to make the constructor public and force the caller to use `makeUniqueWithoutRefCountedCheck` (without a create) function. Otherwise, this is too error-prone now that we're ref-counted.

RupinMittal replied:
You're right, it failed to build. I also had to make the inheritance from IPC::MessageReciever be public instead of private. It should build now. I also removed the create function.

cdumez commented on Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetwork.cpp:
```diff
@@ -37,11 +38,29 @@ namespace WebKit {
+#if ENABLE(WEB_RTC)
+    : m_mdnsRegister(*this)
+#endif
+    , m_webProcess(webProcess)
```
This won't build if `ENABLE(WEB_RTC)` is false.

cdumez replied:
Just make `m_webProcess` the first data member to avoid the issue I guess.

RupinMittal replied:
Oh that's a good point. Definitely saved me from waiting for the bots. Thank you and fixed :)

cdumez commented on Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetwork.h:
```diff
@@ -35,15 +35,20 @@
     WTF_MAKE_TZONE_ALLOCATED(LibWebRTCNetwork);
 public:
-    static UniqueRef<LibWebRTCNetwork> create() { return UniqueRef { *new LibWebRTCNetwork() }; }
+    LibWebRTCNetwork(WebProcess&);
```
`explicit`

cdumez commented on Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetwork.h:
```diff
@@ -35,15 +35,20 @@
+class WebProcess;
 
-class LibWebRTCNetwork : private FunctionDispatcher, private IPC::MessageReceiver {
+class LibWebRTCNetwork : private FunctionDispatcher, public IPC::MessageReceiver {
```
I think we can mark the class as `final`

cdumez commented on Source/WebKit/WebProcess/Network/webrtc/WebMDNSRegister.h:
```diff
@@ -30,31 +30,27 @@
 class WebMDNSRegister : public CanMakeWeakPtr<WebMDNSRegister> {
 public:
-    WebMDNSRegister() = default;
+    WebMDNSRegister(LibWebRTCNetwork&);
```
`explicit`

cdumez reviewed (approved):
LGTM

RupinMittal committed:
Drop IsDeprecatedWeakRefSmartPointerException from WebMDNSRegister
https://bugs.webkit.org/show_bug.cgi?id=281729
rdar://138170193

Reviewed by Chris Dumez.

WebMDNSRegister forwards its ref-counting to its owner LibWebRTCNetwork,
which forwards its own ref-counting to its owner WebProcess.

* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetwork.cpp:
(WebKit::LibWebRTCNetwork::LibWebRTCNetwork):
(WebKit::LibWebRTCNetwork::ref const):
(WebKit::LibWebRTCNetwork::deref const):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetwork.h:
(WebKit::LibWebRTCNetwork::create): Deleted.
(WebKit::LibWebRTCNetwork::connection): Deleted.
(WebKit::LibWebRTCNetwork::isActive const): Deleted.
(WebKit::LibWebRTCNetwork::monitor): Deleted.
(WebKit::LibWebRTCNetwork::socketFactory): Deleted.
(WebKit::LibWebRTCNetwork::disableNonLocalhostConnections): Deleted.
(WebKit::LibWebRTCNetwork::resolver): Deleted.
(WebKit::LibWebRTCNetwork::mdnsRegister): Deleted.
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.cpp:
(WebKit::LibWebRTCNetworkManager::unregisterMDNSNames):
(WebKit::LibWebRTCNetworkManager::CreateNameForAddress):
* Source/WebKit/WebProcess/Network/webrtc/WebMDNSRegister.cpp:
(WebKit::WebMDNSRegister::WebMDNSRegister):
(WebKit::WebMDNSRegister::ref const):
(WebKit::WebMDNSRegister::deref const):
(WebKit::WebMDNSRegister::registerMDNSName):
* Source/WebKit/WebProcess/Network/webrtc/WebMDNSRegister.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::libWebRTCNetwork):
(WebKit::WebProcess::protectedLibWebRTCNetwork):
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: https://commits.webkit.org/285554@main
